src/renderer/utils/createStorage.ts

```diff
@@ -1,73 +1,72 @@
-// Helper to work with browser's local/session storage api
-
-export interface IStorageHelperOptions {
-  addKeyPrefix?: boolean;
-  useSession?: boolean; // use `sessionStorage` instead of `localStorage`
-}
-
-export function createStorage<T>(key: string, defaultValue?: T, options?: IStorageHelperOptions) {
-  return new StorageHelper(key, defaultValue, options);
-}
-
-export class StorageHelper<T> {
-  static keyPrefix = "lens_";
-
-  static defaultOptions: IStorageHelperOptions = {
-    addKeyPrefix: true,
-    useSession: false,
-  };
-
-  constructor(protected key: string, protected defaultValue?: T, protected options?: IStorageHelperOptions) {
-    this.options = Object.assign({}, StorageHelper.defaultOptions, options);
-
-    if (this.options.addKeyPrefix) {
-      this.key = StorageHelper.keyPrefix + key;
+// Keeps window.localStorage state in external JSON-files.
+// Because app creates random port between restarts => storage session wiped out each time.
+import type { CreateObservableOptions } from "mobx/lib/api/observable";
+
+import path from "path";
+import { app, remote } from "electron";
+import { observable, reaction, when } from "mobx";
+import fse from "fs-extra";
+import { StorageHelper } from "./storageHelper";
+import { clusterStore, getHostedClusterId } from "../../common/cluster-store";
+import logger from "../../main/logger";
+
+let initialized = false;
+const loaded = observable.box(false);
+const storage = observable.map<string/* key */, any /* serializable */>();
+
+export function createStorage<T>(key: string, defaultValue?: T, observableOptions?: CreateObservableOptions) {
+  const clusterId = getHostedClusterId();
+  const savingFolder = path.resolve((app || remote.app).getPath("userData"), "lens-local-storage");
+  const jsonFilePath = path.resolve(savingFolder, `${clusterId ?? "app"}.json`);
+
+  if (!initialized) {
+    initialized = true;
+
+    // read once per cluster domain
+    fse.readJson(jsonFilePath)
+      .then((data = {}) => storage.merge(data))
+      .catch(() => null) // ignore empty / non-existing / invalid json files
+      .finally(() => loaded.set(true));
+
+    // bind auto-saving
+    reaction(() => storage.toJSON(), saveFile, { delay: 250 });
+
+    // remove json-file when cluster deleted
+    if (clusterId !== undefined) {
+      when(() => clusterStore.removedClusters.has(clusterId)).then(removeFile);
     }
   }
 
-  protected get storage() {
-    if (this.options.useSession) return window.sessionStorage;
-
-    return window.localStorage;
-  }
-
-  get(): T {
-    const strValue = this.storage.getItem(this.key);
-
-    if (strValue != null) {
-      try {
-        return JSON.parse(strValue);
-      } catch (e) {
-        console.error(`Parsing json failed for pair: ${this.key}=${strValue}`);
-      }
+  async function saveFile(json = {}) {
```

Comment:
Maybe `saveToAppFiles()` could be used instead? `src/common/utils/saveToAppFiles.ts`

Reply:
Maybe, but it uses `fs-extra` sync file operations, e.g. `fse.writeFileSync`